File: typescript/sudoku.js
"use strict";
exports.__esModule = true;
var fs = require("fs");
function string_to_puzzle(input) {
    return input.split('').map(function (c) { return c === '.' ? 0 : parseInt(c, 10); });
}
function puzzle_to_string(puzzle) {
    return puzzle.map(function (e) { return e === 0 ? '.' : e.toString(); }).join('');
}
function is_valid_row(row) {
    var checkset = new Array(10).fill(0);
    for (var _i = 0, row_1 = row; _i < row_1.length; _i++) {
        var e = row_1[_i];
        checkset[e] += 1;
    }
    return checkset.slice(1).every(function (e) { return e <= 1; });
}
function is_valid(grid) {
    for (var r = 0; r < 9; r++) {
        var row = grid.slice(r * 9, r * 9 + 9);
        if (!is_valid_row(row)) {
            return false;
        }
    }
    for (var c = 0; c < 9; c++) {
        var col = [];
        for (var r = 0; r < 9; r++) {
            col.push(grid[r * 9 + c]);
        }
        if (!is_valid_row(col)) {
            return false;
        }
    }
    for (var i = 0; i < 9; i++) {
        var box_row = Math.floor(i / 3);
        var box_col = i % 3;
        var box = [];
        for (var r = box_row * 3; r < box_row * 3 + 3; r++) {
            for (var c = box_col * 3; c < box_col * 3 + 3; c++) {
                box.push(grid[r * 9 + c]);
            }
        }
        if (!is_valid_row(box)) {
            return false;
        }
    }
    return true;
}
function next_open(grid) {
    for (var i = 0; i < 81; i++) {
        if (grid[i] === 0) {
            return i;
        }
    }
    return -1;
}
function solve(grid) {
    if (!is_valid(grid)) {
        return;
    }
    var p = next_open(grid);
    if (p < 0) {
        return grid;
    }
    for (var v = 1; v <= 9; v++) {
        grid[p] = v;
        var result = solve(grid);
        if (result) {
            return result;
        }
        grid[p] = 0;
    }
    return;
}
function main() {
    if (process.argv.length < 2) {
        console.log("Please provide the name of an input file as an argument.");
        process.exit(1);
    }
    // Get the filename from the first command line argument
    var filename = process.argv[2];
    // Read the file line by line
    var lines = fs.readFileSync(filename, 'utf8')
        .split('\n')
        .map(function (e) { return e.trim(); })
        .filter(function (e) { return e.length > 0; });
    // Print each line
    for (var i = 0; i < lines.length; i += 2) {
        var input = lines[i];
        var expected = lines[i + 1];
        var start = performance.now();
        var puzzle = string_to_puzzle(input);
        solve(puzzle);
        var output = puzzle_to_string(puzzle);
        var end = performance.now();
        var duration = end - start;
        if (expected === output) {
            console.log("Solved sudoku ".concat(input, " in ").concat(duration, " ms"));
        }
        else {
            console.log("Failed to solve sudoku ".concat(input, ". Expected ").concat(expected, ", got ").concat(output, "s"));
        }
    }
}
if (require.main === module) {
    main();
}

File: lua/sudoku.lua
function string_to_puzzle(input)
    puzzle = {}
    for i = 1, #input do
        if input:sub(i, i) ~= '.' then
            table.insert(puzzle, tonumber(input:sub(i, i)))
        else
            table.insert(puzzle, 0)
        end
    end
    return puzzle
end

function puzzle_to_string(puzzle)
    local s = {}
    for e in puzzle do
        if e ~= 0 then
            table.insert(s, tostring(e))
        else
            table.insert(s, '.')
        end
    end
    return table.concat(s, '')
end

function is_valid_row(row)
    local checkset = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0}
    for _, e in ipairs(row) do
        if e >= 1
            checkset[e] = checkset[e] + 1
        end
    end
    for _, e in ipairs(checkset) do
        if e > 1 then
            return false
        end
    end
    return true
end

function is_valid(grid)
    for r = 0, 8 do
        row = {0, 0, 0, 0, 0, 0, 0, 0, 0}
        for c = 0, 8 do
            row[c+1] = grid[r*9+c+1]
        end
        if not is_valid_row(row) then
            return false
        end
    end
    for c = 0, 8 do
        col = {0, 0, 0, 0, 0, 0, 0, 0, 0}
        for r = 0, 8 do
            col[r+1] = grid[r*9+c+1]
        end
        if not is_valid_row(col) then
            return false
        end
    end
    for i = 0, 8 do
        box_row = math.floor(i / 3)
        box_col = i % 3
        box = {0, 0, 0, 0, 0, 0, 0, 0, 0}
        k = 1
        for r = box_row*3, box_row*3+2 do
            for c = box_col*3, box_col*3+2 do
                box[k+1] = grid[r*9+c+1]
                k = k + 1
            end
        end
        if not is_valid_row(box) then
            return false
        end
    end
    return true
end

function next_open(grid)
    for i = 1, 81 do
        if grid[i] == 0 then
            return i
        end
    end
    return -1
end

function solve(grid)
    if not is_valid(grid) then
        return false
    end
    local p = next_open(grid)
    if p < 0 then
        return true
    end
    for v = 1, 9 do
        grid[p] = v
        if solve(grid) then
            return true
        end
        grid[p] = 0
    end
    return false
end

function read_filename()
    if #arg < 1 then
        print("Please provide the name of an input file as an argument.")
        os.exit(1)
    end
    return arg[1]
end

function read_file(filename)
    lines = {}
    for line in io.lines(filename) do
        cleaned = line:strip()
        if cleaned ~= '' then
            table.insert(lines, cleaned)
        end
    end
    return lines
end

function solve(lines)
    for i = 1, #lines, 2 do
        input = lines[i]
        expected = lines[i+1]
        s = string_to_puzzle(input)
        start = time.time()
        solve(s)
        end = time.time()
        if puzzle_to_string(s) == expected then
            print("Solved sudoku %s in %d ms" % (input, (end-start)*1000))
        else
            print("Failed to solve sudoku %s. Expected %s, got %s" % (input, expected, s))
            exit(1)
        end
    end
end